### ISSUE TITLE
Preparation for Venus OS integration

### ISSUE THREAD

**Issue by mr-manuel** [kwindrem/TailscaleGX#5]
Tailscale will be integrated into Venus OS and be part of the normal and large Venus OS image.

Currently I have no version number, with which the integration will be available, but I will provide it to you as soon as I have it. Instead of a version check, you could do it also by checking if the folder `/opt/victronenergy/tailscale` exists.

The idea for this issue is, that if the SetupHelper recognizes that Tailscale is already present in Venus OS, then it does a migration of the Tailscale settings automatically. After that this plugin is disabled/removed.

## Copy config files
The `tailscaled` state dir is `/data/conf/tailscale`, so you can move all files from `/data/setupOptions/TailscaleGX/state`. Check if that directory already exists before copying it, if not you have to create it.

## Copy and transform dbus settings
`com.victronenergy.settings/Settings/TailscaleGX/IpForwarding` -> `com.victronenergy.settings/Settings/Services/Tailscale/CustomArguments`
If `0` a transformation is not needed, if `1` set the value to `--advertise-exit-node=true`.

`com.victronenergy.settings/Settings/TailscaleGX/Enabled` -> `com.victronenergy.settings/Settings/Services/Tailscale/Enabled`.

If you have any questions let me know.

**Comment by kwindrem** [comment]
Thanks for the heads up. I'll make the suggested changes soon.

**Comment by mr-manuel** [comment]
You're welcome. I don't expect it to get released before the middle of next week, so it's not that urgent.

**Comment by kwindrem** [comment]
seems like com.victronenergy.settings/Settings/Services/Tailscale/CustomArguments might include other arguments.

Should I be appending --advertise-exit-node=true to it rather than replacing it’s value?
—
Kevin

**Comment by kwindrem** [comment]
After additional thought I decided if the new setting exists I won’t touch it so won’t need to append to an existing string.
—
Kevin

**Comment by mr-manuel** [comment]
This value will be empty, since on the first time it's present, the migration will happen.

**Comment by kwindrem** [comment]
I'm having trouble setting .../CustomArgument to --advertise-exit-node=true

This is in a bash script which uses the dbus -y ... SetValue call

I am getting dbus: error: no such option: --advertise-exit-node

**Comment by mr-manuel** [comment]
To ensure that `--advertise-exit-node` is interpreted as a value rather than a command-line option, you can use the `--` separator, which tells the command-line parser to treat all subsequent arguments as positional parameters rather than options.

```bash
dbus -y com.victronenergy.settings /Settings/Services/Tailscale/CustomArguments SetValue -- "--advertise-exit-node"
```

By adding `--` before `--advertise-exit-node`, you ensure that it is treated as a value rather than an option.

**Comment by kwindrem** [comment]
Thanks. That worked

**Comment by kwindrem** [comment]
I've made modifications to TailscaleGX as you suggested but I do have a question regarding migration:

Will the new settings and state already exist in the stock firmware before TailscaleGX or it's setup script have a chance to run?

The reason I ask is my code currently checks for the presence of the stock items and does not migrate from the TailscaleGX ones if they do exist. This would mean tailscale would start up in its default configuration and establishing a connection again. Maybe this would be OK and I don't need to migrate from TailscaleGX.

Comments?

**Comment by mr-manuel** [comment]
The new settings and state will exists before `/data/rc*.local` run, therefore no chance that the setup script runs before.

**Comment by kwindrem** [comment]
I think it may be best then to not attempt any migration. This means the user/installer would need to set up tailscale again but with the QR code this should be fairly painless.

Comments?

**Comment by kwindrem** [comment]
An alternative would be to update TailscaleGX to migrate to the new settings/state now so they are using them when the firmware includes tailscale natively.

**Comment by mr-manuel** [comment]
That was the idea, that you update TailscaleGX now to allow a seamless migration.

**Comment by kwindrem** [comment]
TailscaleGX has been updated to use the new Settings parameters and state location.

Running TailscaleGX-control will migrate the changes, then if /opt/v*/tailscale is detected, it will exit.

Running the setup script if /opt/v*/tailscale exists will fail to install.

When the firmware version that includes tailscale is determined, I'll update TailscaleGX so an install won't even be possible (obsolete version will be set).

These changes are on the latest tag and are currently v1.10. Feel free to test.

**Comment by kwindrem** [comment]
What is the current state of tailscale being incorporated into Venus OS. I recently had someone ask on the Victron forum if it’s implemented and I did find menus under services. The enable on/off is hooked up to the dbus parameter. However the menu just says Starting…

I looked and there is no /opt/victronenerfy/tailscale directory so my setup script continues with installation of TailscaleGX.

I saw a comment somewhere from mvader indicating that it’s all working but can’t find any release notes about it.

--
Kevin Windrem
***@***.***

530-913-5659

**Comment by mr-manuel** [comment]
I can give you an update at the end of this week. We made a huge progress with the ressource limitation of Tailscale, since if a user transfers a lot of data over a GX device, it can max out the CPU and the whole power system becomes unstable.

**Comment by kwindrem** [comment]
OK. Just let me know when it's implemented and I'll test my script to make sure TailscaleGX uninstalls properly with the native tailscale. I'll also set an incompatible version in TailscaleGX to make sure it isn't allowed to install starting with that version.

**Comment by duemme** [comment]
Any update on when tailscale will be available on Venus OS? Thanks

**Comment by kwindrem** [comment]
Nope

**Comment by mr-manuel** [comment]
It should yes. Testing will be only possible shortly before release.

Anyway if you have enabled the access via VRM portal, then you always can connect via VRM and setup Tailscale without the need of an SSH connection.

**Comment by kwindrem** [comment]
Some time ago, I moved TailscaleGX config files to match where the Victoron implementation will place them. So far as these files are compatible with the Tailscale version implemented as part of Venus OS there should be no interruption of services. We will not know for sure until the Venus OS version implementing Tailscale natively is released however.

**Comment by kwindrem** [comment]
I recently had a request to implement tailscale auth keys in TailscaleGX. I have this working however the GUI (v1) does not permit the long key to be entered. The key can be set via dbus-spy and all works well.

Not sure if you want to include auth keys in the stock tailscale implementation.

**Comment by mr-manuel** [comment]
I don't think that is needed, since it's a very edge case. If the user wants to have it anyway there is already a field for custom commands, so the user can add it there. Thanks for the hint!

**Comment by kwindrem** [comment]
Victron has indicated they are delaying native tailscale support in Venus OS due to potential security issues.

TailscaleGX will continue to be supported with configuration through gui-v1
(not the New UI (aka gui-v2). You can also run /data/TailscaleGX/tailscale for configuration and use dbus-spy to enable it.
